File: templates/hooks/opencode/dream-wf-guard.js
import fs from 'node:fs';
import path from 'node:path';

const mutatingTools = new Set(['Write', 'Edit', 'MultiEdit', 'Delete', 'apply_patch', 'shell']);
const mutatingShell = /\b(rm|mv|cp|mkdir|touch|npm\s+install|pnpm\s+add|yarn\s+add|bun\s+add|git\s+commit|git\s+push)\b/;

export default function dreamWfGuard() {
  return {
    name: 'dream-wf-guard',
    async 'tool.execute.before'(input) {
      // 逃生舱：DREAM_WF_MODE=advisory 时跳过 strict 检查。
      if (process.env.DREAM_WF_MODE?.toLowerCase() === 'advisory') {
        return;
      }

      const toolName = input?.tool || input?.tool_name || '';
      const toolInput = input?.input || input?.tool_input || {};
      const cwd = input?.cwd || process.cwd();
      const root = findRoot(cwd);
      const command = typeof toolInput.command === 'string' ? toolInput.command : '';
      const isMutating = mutatingTools.has(toolName) || ((toolName === 'Shell' || toolName === 'Bash') && mutatingShell.test(command));

      if (!isMutating) {
        return;
      }

      const tasks = activeTasks(root);
      if (tasks.length === 0) {
        throw new Error('dream-wf strict: mutating actions require an active Trellis task. Create or start a Trellis task first, or switch dream-wf to advisory mode (DREAM_WF_MODE=advisory).');
      }

      // 规划产物始终允许，便于在 planning 阶段编写 prd/design 等。
      if (isPlanningArtifact(root, toolInput)) {
        return;
      }

      // 若存在已确认的 in_progress 任务，允许实现操作，不被其它 stale planning 任务阻塞。
      const hasConfirmedInProgress = tasks.some(({ taskDir, task }) => task.status === 'in_progress' && isPrdConfirmed(taskDir));
      if (hasConfirmedInProgress) {
        return;
      }

      const hasAnyInProgress = tasks.some(({ task }) => task.status === 'in_progress');
      if (hasAnyInProgress) {
        return;
      }

      // 剩余情况：所有活跃任务都是 planning。若任一未确认 PRD，则阻塞实现。
      const hasUnconfirmedPlanning = tasks.some(({ taskDir, task }) => task.status === 'planning' && !isPrdConfirmed(taskDir));
      if (hasUnconfirmedPlanning) {
        throw new Error('dream-wf strict: implementation is blocked while all active tasks are in planning and at least one PRD is not confirmed. Continue grill-me PRD clarification first. Planning artifacts under .trellis/tasks/** are allowed.');
      }
    }
  };
}

function findRoot(cwd) {
  let current = path.resolve(cwd);
  while (true) {
    if (fs.existsSync(path.join(current, '.trellis'))) {
      return current;
    }
    const parent = path.dirname(current);
    if (parent === current) {
      return path.resolve(cwd);
    }
    current = parent;
  }
}

function activeTasks(root) {
  const tasksDir = path.join(root, '.trellis', 'tasks');
  if (!fs.existsSync(tasksDir)) {
    return [];
  }

  return fs.readdirSync(tasksDir, { withFileTypes: true })
    .filter((entry) => entry.isDirectory())
    .map((entry) => {
      const taskDir = path.join(tasksDir, entry.name);
      const taskJson = path.join(taskDir, 'task.json');
      if (!fs.existsSync(taskJson)) {
        return undefined;
      }
      try {
        const task = JSON.parse(fs.readFileSync(taskJson, 'utf8'));
        return { taskDir, task };
      } catch {
        return undefined;
      }
    })
    .filter(Boolean)
    .filter(({ task }) => task.status === 'planning' || task.status === 'in_progress');
}

function isPrdConfirmed(taskDir) {
  const prd = path.join(taskDir, 'prd.md');
  if (!fs.existsSync(prd)) {
    return false;
  }
  const text = fs.readFileSync(prd, 'utf8').toLowerCase();
  return ['prd confirmed', 'confirmed: true', 'status: confirmed', '用户已确认', '已确认'].some((marker) => text.includes(marker));
}

function isPlanningArtifact(root, toolInput) {
  const candidate = toolInput?.path || toolInput?.file_path || toolInput?.target_file || '';
  if (!candidate) {
    return false;
  }

  let relative;
  try {
    const rootResolved = fs.realpathSync(root);
    const filePath = path.isAbsolute(candidate) ? candidate : path.resolve(rootResolved, candidate);
    relative = path.relative(rootResolved, fs.realpathSync(filePath)).split(path.sep).join('/');
  } catch {
    return false;
  }

  if (!relative.startsWith('.trellis/tasks/')) {
    return false;
  }

  const name = path.basename(relative);
  return ['prd.md', 'design.md', 'implement.md', 'implement.jsonl', 'check.jsonl'].includes(name) || relative.includes('/research/');
}

File: templates/hooks/codex/dream-wf-guard.py
#!/usr/bin/env python3
# Codex PreToolUse guard。
# Codex CLI 支持 PreToolUse 阻塞式 hook，配置在 .codex/hooks.json。
# stdin 收 JSON payload，stdout 返回 permissionDecision，或 exit code 2 表示 block。

import json
import os
import re
import sys
from pathlib import Path

# Codex 的变更类 tool name：shell、apply_patch（文件编辑）。
MUTATING_TOOLS = {"Write", "Edit", "MultiEdit", "Delete", "apply_patch", "shell", "Bash"}
MUTATING_SHELL = re.compile(r"\b(rm|mv|cp|mkdir|touch|npm\s+install|pnpm\s+add|yarn\s+add|bun\s+add|git\s+commit|git\s+push)\b")


def find_root(cwd):
    current = Path(cwd).resolve()
    for candidate in [current, *current.parents]:
        if (candidate / ".trellis").exists():
            return candidate
    return current


def active_tasks(root):
    tasks_dir = root / ".trellis" / "tasks"
    if not tasks_dir.exists():
        return []

    result = []
    for task_json in tasks_dir.glob("*/task.json"):
        try:
            data = json.loads(task_json.read_text())
        except Exception:
            continue
        if data.get("status") in {"planning", "in_progress"}:
            result.append((task_json.parent, data))
    return result


def is_prd_confirmed(task_dir):
    prd = task_dir / "prd.md"
    if not prd.exists():
        return False
    text = prd.read_text(errors="ignore").lower()
    markers = ["prd confirmed", "confirmed: true", "status: confirmed", "用户已确认", "已确认"]
    return any(marker in text for marker in markers)


def is_planning_artifact(root, tool_input):
    if not isinstance(tool_input, dict):
        return False

    candidate = tool_input.get("path") or tool_input.get("file_path") or tool_input.get("target_file") or ""
    if not candidate:
        return False

    try:
        root_resolved = root.resolve()
        file_path = Path(candidate)
        if not file_path.is_absolute():
            file_path = (root_resolved / file_path)
        relative = file_path.resolve().relative_to(root_resolved).as_posix()
    except Exception:
        return False

    if not relative.startswith(".trellis/tasks/"):
        return False

    name = Path(relative).name
    return (
        name in {"prd.md", "design.md", "implement.md", "implement.jsonl", "check.jsonl"}
        or "/research/" in relative
    )


def deny(message):
    print(json.dumps({
        "hookSpecificOutput": {
            "hookEventName": "PreToolUse",
            "permissionDecision": "deny",
            "permissionDecisionReason": message
        }
    }))
    sys.exit(0)


def allow():
    print(json.dumps({
        "hookSpecificOutput": {
            "hookEventName": "PreToolUse",
            "permissionDecision": "allow"
        }
    }))
    sys.exit(0)


def main():
    # 逃生舱：DREAM_WF_MODE=advisory 时跳过 strict 检查。
    if os.environ.get("DREAM_WF_MODE", "").lower() == "advisory":
        allow()

    try:
        payload = json.load(sys.stdin)
    except Exception:
        allow()

    tool_name = payload.get("tool_name") or payload.get("tool", "")
    tool_input = payload.get("tool_input") or payload.get("input") or {}
    cwd = payload.get("cwd") or os.environ.get("CODEX_PROJECT_DIR") or os.getcwd()
    root = find_root(cwd)

    is_mutating = tool_name in MUTATING_TOOLS
    command = tool_input.get("command", "") if isinstance(tool_input, dict) else ""
    if tool_name in {"Shell", "Bash", "shell"} and MUTATING_SHELL.search(command):
        is_mutating = True

    if not is_mutating:
        allow()

    tasks = active_tasks(root)
    if not tasks:
        deny("dream-wf strict: mutating actions require an active Trellis task. Create or start a Trellis task first, or switch dream-wf to advisory mode (DREAM_WF_MODE=advisory).")

    if is_planning_artifact(root, tool_input):
        allow()

    in_progress_confirmed = any(
        task.get("status") == "in_progress" and is_prd_confirmed(task_dir)
        for task_dir, task in tasks
    )
    if in_progress_confirmed:
        allow()

    in_progress_any = any(task.get("status") == "in_progress" for _, task in tasks)
    if in_progress_any:
        allow()

    planning_unconfirmed = [
        task_dir for task_dir, task in tasks
        if task.get("status") == "planning" and not is_prd_confirmed(task_dir)
    ]
    if planning_unconfirmed:
        deny("dream-wf strict: implementation is blocked while all active tasks are in planning and at least one PRD is not confirmed. Continue grill-me PRD clarification first. Planning artifacts under .trellis/tasks/** are allowed.")

    allow()


if __name__ == "__main__":
    main()
